### ISSUE TITLE
IBM 8514/A Emulation fails with HDIDEMO.exe (IBM 8514/A Demo Program)

### ISSUE THREAD

**Issue by matt713** [86Box/86Box#2968]
**Describe the bug**
When run "HDIDEMO.EXE" (IBM 8514/A Adapter Interface Demonstration program for DOS), the rendered picture is corrupted (screenshot is attached), the demo program doesn't complete and it crashes, i.e. it doesn't response to keyboard input any longer.

**To Reproduce**
1. Setup Machine (PS2 model with MCA slots or PC with ISA slots) in 86Box with "8514/A" option checked - I've tried many 86Box Machines and it doesn't matter for the problem, that's why don't provide more specific details 

2. Install DOS (I used IBM DOS 6.30)

3. Download IBM diskette image of their "Interface Code for 8514/A v1.02" - Please, note that IBM provided that publicly on their FTP server at URL: "ftp://ftp.pc.ibm.com/pub/pccbbs/refdisks/interfac.exe" before the FTP was shutdown in 1999. So, the file is available all over the Internet on Mirrors/Archives of the old IBM FTP server. I got it from here:

http://www.walshcomptech.com/selectpccbbs/interfac.zip

(BTW, v1.02 is the latest version, but I've tested with v1.00 as well - it doesn't matter for the problem.)

4. load the diskette image from step 3, go to drive a: and type "install a: c:", follow the on-screen instructions, skip installing OS2 files, as we're using DOS in this particular case

5. after the installation is complete:
> cd c:\hdipcdos
> hdiload.exe ;IBM TSR program for DOS that provides "8514A Adapter Interface" for the application software
> hdidemo.exe ;IBM 8514/A Adapter Interface Demo program for DOS

**Expected behavior**
Proper picture rendered from HDIDEMO.exe without any corruption, no crash and it's responding to keyboard input, i.e. pressing "Enter" exits the Demo.

**Screenshots**
![hdidemo_crash](https://user-images.githubusercontent.com/59395613/207840361-61cf5180-c3b7-4036-9f1f-c786d6dfa8e2.png)

**Desktop (please complete the following information):**
 - OS: Win 7 x64
 - 86Box version: tested on "3.11 build 4311"
 - Build information: Regular build, AMD64 Release 3.11 version

**Comment by TC1995** [comment]
let me check the accelerated commands...

**Comment by matt713** [comment]
> let me check the accelerated commands...

Thank you and I forgot to mention, the full C source code of HDIDEMO, "HDIDEMO.C", is included in the IBM diskette image I mentioned in step 3 of "how to reproduce" section in my report.

**Comment by TC1995** [comment]
Well, does it also happen in the 32-bit version of 86box?

**Comment by matt713** [comment]
> Well, does it also happen in the 32-bit version of 86box?

I am going to test right now and report back as soon as I am finished...

**Comment by matt713** [comment]
> > Well, does it also happen in the 32-bit version of 86box?
> 
> I am going to test right now and report back as soon as I am finished...

using 32-bit of 86box is actually worse - attached screenshot.
![hdidemo_bad_32bit_86box](https://user-images.githubusercontent.com/59395613/207847720-0e70b467-4cf2-406b-854d-824ea855648d.png)

**Comment by matt713** [comment]
@TC1995 : I saw you're the author of both 8514A and XGA emulation. So, it's completely off-topic, but just to use the opportunity I have contact with you here, because I am not sure of the XGA emulation status and if it's a bug or just the current status of the XGA emulation, but IBM XGADEMO fails completely to detect the XGA adapter in DOS, different versions of the IBM XGA demo are available here:

http://www.walshcomptech.com/selectpccbbs/xga210.zip
http://www.walshcomptech.com/selectpccbbs/xga211.zip
http://www.walshcomptech.com/selectpccbbs/xga212.zip

It's highly interrelated with 8514A, because the Adapter Interface for XGA includes as subset the Adapter Interface for 8514A. Anyway, that's just to get those XGA Demos if you don't have them and have not tested the emulation against them.

**Comment by TC1995** [comment]
I already know about the XGA one, which uses some undocumented parts.

**Comment by matt713** [comment]
> I already know about the XGA one, which uses some undocumented parts.

at least all AI functions: the 8514A Base Functions Subset and the 2 Extension Sets (XGA-1 and XGA-2) are documented in "IBM PS2 XGA Adapter Interface Technical Reference":

https://archive.org/details/bitsavers_ibmpcps215nterfaceTechnicalReferenceSep90_5765444

however, I noticed some inconsistencies, between "8514A Base Functions Subset" based on IBM header file (AFIDATA.H) from v1.x release for 8514A and the one from v2.x release for XGA and the comprehensive documentation of the same fucntions included in "Programmer's Guide To The EGA, VGA, And Super VGA Cards (3rd Edition)" :

https://archive.org/details/programmersguidetotheegavgaandsupervgacardsbyrichardf.ferraro19943rdedition

starting on page 928, but at least there several sources and hundreds of pages where those AI fictions are documented - it's still much better than noting.

**Comment by matt713** [comment]
on-topic update: I made 86box PC Machine and choose "Tseng Labs ET4000w32p Rev. D (Diamond Stealth32)" for video card, then loaded 8 Planes (i.e. 1Mb, since 1 Plane of 8514A memory is 1024 bits) of Tseng 8514A Emulation for DOS:

https://support.ptc.com/support/cs_cv/support/pd_video_vesa_drivers/8514A.ZIP

![00_Tseng Labs ET4000w32p Rev  D (Diamond Stealth32)_8514A_Emu](https://user-images.githubusercontent.com/59395613/207925263-adefc2cb-fbab-4189-adbc-a53ce437e27c.png)

then IBM HDI Demo runs perfectly. So, now we know how IBM 8514A Demo is supposed to look like - it's kind of Animated, i.e. the "fins" colors are changing, but more careful observations gives they are rotating in clockwise direction. I am attaching screenshots:
![01_ibm_hdi_demo_tseng4000](https://user-images.githubusercontent.com/59395613/207925432-dd1af35a-d068-4b82-9c47-fc92bed4bc70.png)
![02_ibm_hdi_demo_tseng4000](https://user-images.githubusercontent.com/59395613/207925446-d3f73604-d109-428d-93a5-b9a88da7fea9.png)
![03_ibm_hdi_demo_tseng4000](https://user-images.githubusercontent.com/59395613/207925460-750a936c-1bf5-4bd8-a469-9d6a45acdc5f.png)
![04_ibm_hdi_demo_tseng4000](https://user-images.githubusercontent.com/59395613/207925468-d0471938-4575-4569-b187-ce268db39808.png)

**Comment by TC1995** [comment]
It's more about the emulation in the XGA core of the emulator.

**Comment by matt713** [comment]
> It's more about the emulation in the XGA core of the emulator.

even currently buggy, IMHO your 8514A emulator is tremendous progress in the right direction towards working 8514A emulation. in the meantime, it's a bonus, the 8514A emulation for Tseng4000 is working and can be used for reference what is supposed to be the expected output from a 8514A application.

however, currently with the XGA emulation in 86box, at least I don't get any progress - I mean IBM XGA Demo program cannot even see the XGA card present, but IBM "bios" (even in PS2 world I believe that's not the right word, how it's called - "IBM reference disk") sees the XGA card:
![00_xga](https://user-images.githubusercontent.com/59395613/207928934-7074ee10-d69d-4baa-ad65-30770f4d574d.png)
![01_xga](https://user-images.githubusercontent.com/59395613/207928947-a6b3d070-e5de-4d57-926d-8acdc511af73.png)

but even the "XGAAIDOS.SYS" is loaded, i.e. the AI for XGA, the IBM XGA Demo program complains it's not available:

![02_xga](https://user-images.githubusercontent.com/59395613/207930375-7de0368e-b137-4aaf-b307-f119c8929696.png)

**Comment by TC1995** [comment]
I already know about the last screenshot you've posted but I don't know how to solve that.

**Comment by matt713** [comment]
> I already know about the last screenshot you've posted but I don't know how to solve that.

no worries, I just posted it for completeness here for other people, because both 8514A and XGA emulations are new and I think most people are unaware of their current status. Let's hope more people will be interested and someone will be actually able to help.

**Comment by matt713** [comment]
BTW, do you need BIOS dump(s) of ATI Mach32 - it's PCI card 100% Register-Level compatible with IBM 8514A? So, you can add Mach32 to the 8514A emulator. Basically, we can MCA (IBM original card), ISA (several clones) and PCI card (ATI Mach32) that are 8514A cards.

**Comment by OBattler** [comment]
Yes, we do.

**Comment by matt713** [comment]
> Yes, we do.

OK, here is the first one:

[https://www.mediafire.com/file/27u399ocdpvc890/01_mach32_rom.zip/file](url)

it's from a Mach32 PCI card I repaired and that's why the BIOS dump is made with hardware programmer. In the zip I included TXT with board part number, BIOS sticker markings and few major chips markings. I have at least 1 or 2 more Mach32 PCI cards with different BIOS versions that I can dump and post here in the next few days.

**Comment by matt713** [comment]
short off-topic again about XGA - as far as I understand the only ISA-clones are based around IMSG200 chip. actually not technically clones as they obtained officially license from IBM - even the chip has "XGA(TM)" marked on it. 

IMSG200 datasheet is available on many places, for example here:

https://ardent-tool.com/datasheets/INMOS_IMSG200.pdf

and recently the BIOS of such ISA card was dumped:

https://www.vogons.org/viewtopic.php?f=63&t=74694

So, maybe, XGA can be attacked from another side - IMSG200 card emulator, since it's licensed by IBM, then maybe the issue with XGA Demo and AI (Adapter Interface) will be resolved, as currently @TC1995 has no idea what is the cause.

**Comment by matt713** [comment]
made bios dump of another different PCI Mach32 card, here it is:

https://www.mediafire.com/file/m4r5lqpg9jt3dq1/mach32_pci_vbios_02.zip/file

**Comment by matt713** [comment]
I should have opened new issue where to discuss XGA, because I already made too many off-topic entries about it here, but I guess it's too late now and better everything be in the same place. So, I believe I am on a "hot"-lead why XGADEMO fails and basically says "XGAAIDOS.SYS" is not loaded, even it's loaded. So, when I run "XGA-2 System Information and Utilities", there is line that says:

`Checking for XGA DMQS BIOS interface
DMQS not supported`

but it seems "XGAAIDOS.SYS" uses/needs "DMQS", because the line in CONFIG.SYS is:

`DEVICE=C:\XGAPCDOS\XGAAIDOS.SYS C:\XGA$DMQS\`

but I don't have "XGA$DMQS" created from the installer - my best guess is because it detects as DMQS is not supported. I don't know maybe it's a wrong lead, but maybe "XGAAIDOS.SYS" doesn't work properly without DMQS and that's the root cause.

**Comment by matt713** [comment]
found 3rd Mach32 PCI card in my storage and dumped the vBIOS, it's here:
https://www.mediafire.com/file/f2oyq6t72yjh73c/mach32_pci_vbios_03.zip/file

compared to the previous Mach32 PCI cards of which I made ROM dumps:
https://github.com/86Box/86Box/issues/2968#issuecomment-1353613733
https://github.com/86Box/86Box/issues/2968#issuecomment-1356170091

this one is very interesting, because it has 2MB RAM onboard, i.e. soldered and not upgradable or downgradable like more common Mach32 PCI cards, which have 1MB soldered and 8 empty sockets to upgrade them to 2MB. 

Also, it has "ATI ATI68875" RAMDAC, which is very rare and one of the few:

https://gona.mactar.hu/DOS_TESTS/ramdac.html

that work properly in some special occasions. Last, but not least, I don't know if all 2MB are utilized for 8514A, but ATI made Technical note that:

> ATI HDILOAD.EXE, version 2.3 or higher, otherwise with 2MB of video RAM, 
> the application returns an error saying there is not enough video RAM to support the operation

Now, ATI HDILOAD v2.1a is very easy to find, but ATI HDILOAD v2.3 is very rare, I found it here:

http://vgamuseum.info/images/doc/ati/mach32driver.zip

it's located on "Disk3" in archive "AI.LZH"

and just for completeness ATI HDILOAD v2.1a link:

https://ardent-tool.com/video/hdload.zip

and last note - there is full package of ATI HDILOAD v2.1a as well, i.e. containing drivers for several OS, etc. but the above link is enough as far as 8514/A is concerned.

**Comment by matt713** [comment]
I've done some more experiments, but this time with simple test program (it draws just 4 small squares) and it seems the root cause of all problems starts at the very beginning with the call to HOPEN() - my best guess is some buffer(s) and/or initialization data are set wrong in 'vid_8514a.c' based on what parameters are passed via the call to HOPEN() and thus the screen is not redrawn properly. So, here are the further details:

```
typedef struct {
   word   length;
   byte   oflags;
   byte   mode;
   byte   iflags;
} HOPEN_DATA;
```

where the meaning of 'mode' is as follows:

> 0x00: 1024x768, Text:    85x38
> 0x01:     640x480, Text:    80x34
> 0x02: 1024x768, Text: 128x54
> (0x03: 1024x768, Text: 146x51, not tested by me)

So, the call to 'HOPEN' is what triggers the problems:

```
openData.mode = 0; //always fails, very bad, keeps the previously typed text on the command prompt
HOPEN(&openData);
```

```
openData.mode = 1; //works just fine
HOPEN(&openData);
```

```
openData.mode = 2; //random or unstable behavior  - one time it works, one time it fails
HOPEN(&openData);
```

I am attaching screenshots showing how the above tests output looks like exactly. So, basically it seems 'mode = 1' is the only properly working one, which explains why IBM demo program fails - it uses the default 'mode = 0'.

=====
mode = 0:

![mode_00_bad](https://user-images.githubusercontent.com/59395613/233674159-1838ee93-7a59-4ad3-8003-5188c12e9bb6.png)

mode = 1:
![mode_01_ok](https://user-images.githubusercontent.com/59395613/233674185-3daf32e7-58a4-40d7-a04f-89d84ce504dd.png)

mode = 2, failed:
![mode_02_failed](https://user-images.githubusercontent.com/59395613/233674208-6d8153a7-2aa3-45c8-9b36-6452db5f4252.png)

mode = 2, succeeded:

![mode_02_ok](https://user-images.githubusercontent.com/59395613/233674596-d08393ba-1694-4605-9964-8d95beac33d9.png)
=====

I am posting this just in case someone familiar with 'vid_8514a.c' has immediate idea based on the above information what is wrong, otherwise most like I will try to debug myself, what is executed differently between 'mode = 1', which is OK and 'mode = 0', which is fails.

**Comment by Notchmaster** [comment]
Sorry for reviving, but testing on the latest experimental build (4694), the HDIDEMO doesnt crash the system anymore, it now moves and animates, and pressing enter ends it. However, the artifacts are still the same.

**Comment by TC1995** [comment]
Need to figure out how the commands display it.

**Comment by matt713** [comment]
>the artifacts are still the same.

IMHO, just from outside observation (by that I mean not looking at the code and try to understand what happens from that side), but just running some simple tests and observe the behavior in practice, the vast majority of artifact seems to me are due to some memory not de-initialized/cleaned and it keeps some previous (bad) state, because the new version (by that I mean the huge code change in "vid_xga.c" after the May 5, 2023 update) compared to the old one version is now free of artifact when the full (whole) screen is drawn, i.e. as I imagine that every portion of the memory is re-initialized, here are 2 images as example for what I mean by that to make it clear:

* **old version full screen - bad, artifacts:**
![old_bad_full_screen](https://github.com/86Box/86Box/assets/59395613/d8872c47-5304-4403-a945-13cbc4cb7ed0)

* **new version full screen - ok, NO artifacts:**
![new_ok_full_screen](https://github.com/86Box/86Box/assets/59395613/bbdddb7b-8199-48c0-a35f-6c1b3bc0d899)

**However, when only potion of the screen is drawn then even with the new code there is no change and it's still exactly what I described here:**

https://github.com/86Box/86Box/issues/2968#issuecomment-1517999773

and more precisely the problem is very obvious with "mode==2":

`openData.mode = 2; //random or unstable behavior  - one time it works, one time it fails`

because in that case it's just random - one time no artifacts, one time artifacts and it's totally random when you get good and when bad image. 

So, as brief summary and to keep it short with the new version now "mode == 1" seems free of artifacts (see the above 2 images I attached, clearly in "mode == 1" as their resolution is 640x480) when the full screen is drawn, but "mode==2" and "mode==0" seem the same as before to me, i.e. as what I described in my previous post that I aforementioned above. 

In any way, @TC1995 made huge progress, but unfortunately it will need more work until it's fully working.
